### PR TITLE
Add an xray meta key/value to the stream data

### LINKED ISSUE #7: Add XRay trace ID to audit-log entries
This is good for tracking which action was caused by what request, if we ever needed to get more data.

### PR DESCRIPTION
**Pull request by pdewouters**
Screenshot 2020-07-06 at 10 06 05

The record is created in `wp_stream_meta`

Fixes #7

### CODE REVIEW

File: inc/stream/namespace.php [hunk 1]
```diff
@@ -8,6 +8,7 @@
 namespace Altis\Security\Stream;
 
 use Altis;
+use function HM\Platform\XRay\get_root_trace_id;
```
Comment (roborourke) [inline inc/stream/namespace.php:11]: We're moving away `use function` statements in favour of using the namespace so it's clearer where a function is from at the point of use, so in this case it'd be `use HM\Platform\XRay` and `XRay\get_root_trace_id()`

File: inc/stream/namespace.php [hunk 2]
```diff
@@ -88,3 +89,15 @@ function override_network_admin_bar_menu( WP_Admin_Bar $wp_admin_bar ) {
 		]
 	);
 }
+
+/**
+ * Add the Xray ID to the log item meta data.
+ *
+ * @param $record
```
Comment (roborourke) [inline inc/stream/namespace.php:96]: Missing param doc, should have a comment

File: inc/stream/namespace.php [hunk 3]
```diff
@@ -88,3 +89,15 @@ function override_network_admin_bar_menu( WP_Admin_Bar $wp_admin_bar ) {
...
+ *
+ * @return mixed
```
Comment (roborourke) [inline inc/stream/namespace.php:98]: The function itself seems to assume that we receive an array and also return array so we should update the type hinting on the function and this `@return` doc accordingly or add handling for any other types of variable that could come through.

**Comment by roborourke** [comment]
@joehoyle is audit log meta the right place to store this or should it be top level?

**Comment by joehoyle** [comment]
@roborourke I'm wondering if it's possible to add it to the top level, when the mysql (or other) drivers are used, as perhaps the top level has to mapped to mysql columns, which we wouldn't want to change. If that's not the case, and Stream has some internal way to handle that, I agree this should be top-level.

**Comment by pdewouters** [comment]
As far as I understand, it does restrict it to the stream database table fields

**Comment by roborourke** [comment]
@joehoyle looks like we might not have anything other than meta that will work in a way that's interoperable with local environment storage. We could update the audit-log plugin to read from the field Paul is setting here in meta at move to the top level at that point if we want to.

File: inc/stream/namespace.php [hunk 4]
```diff
@@ -88,3 +89,15 @@ function override_network_admin_bar_menu( WP_Admin_Bar $wp_admin_bar ) {
...
+
+/**
+ * Add the Xray ID to the log item meta data.
+ *
+ * @param array $record The stream log record to filter.
+ *
+ * @return array
+ */
+function filter_wp_stream_record_array( $record ) : array {
+	$record['meta']['xray'] = XRay\get_root_trace_id();
```
Comment (roborourke) [inline inc/stream/namespace.php:101]: I think XRay can be switched off so you'll need to check the cloud module config here - double check but it'd be something like `Altis\get_config()['modules]['cloud']['xray']`, if that's false then just return the record early.

Comment (joehoyle) [reply inc/stream/namespace.php:101]: You might also want to (or instead) do a `function_exists`, as that would be more reliable

File: inc/stream/namespace.php [hunk 5]
```diff
@@ -88,3 +89,19 @@ function override_network_admin_bar_menu( WP_Admin_Bar $wp_admin_bar ) {
...
+
+/**
+ * Add the Xray ID to the log item meta data.
+ *
+ * @param array $record The stream log record to filter.
+ *
+ * @return array
+ */
+function filter_wp_stream_record_array( $record ) : array {
+	if ( ! function_exists( 'XRay\\get_root_trace_id' ) ) {
```
Comment (roborourke) [inline inc/stream/namespace.php:101]: I think `function_exists()` requires the fully qualified function name eg. `HM\\Platform\\Xray\\get_root_trace_id`. Did you check this with xray switched off via your config? Set the property at the path `extra.altis.modules.cloud.xray` to false in your `composer.json` to confirm it.

Comment (pdewouters) [reply inc/stream/namespace.php]: ah yes good point

File: inc/stream/namespace.php [hunk 6]
```diff
@@ -88,3 +89,19 @@ function override_network_admin_bar_menu( WP_Admin_Bar $wp_admin_bar ) {
...
+
+/**
+ * Add the Xray ID to the log item meta data.
+ *
+ * @param array $record The stream log record to filter.
+ *
+ * @return array
+ */
+function filter_wp_stream_record_array( $record ) : array {
+	if ( ! function_exists( 'HM\\Platform\\Xray\\get_root_trace_id' ) ) {
```
Comment (roborourke) [inline inc/stream/namespace.php:101]:
```suggestion
	if ( ! function_exists( 'HM\\Platform\\XRay\\get_root_trace_id' ) ) {
```

**Pull request merged**